### PR TITLE
Function calling sequence: locals and stack frame

### PR DESCRIPTION
**Pull request by dschuff**

### CODE REVIEW

File: BasicCABI.md [hunk 1]
```diff
@@ -81,3 +81,58 @@ Bitfield type | Witdh *w* | Range
 `int`, `unsigned int` | 1 to 32 | 0 to 2<sup>w</sup>-1
 `signed long long` | 1 to 64 | -2<sup>(w-1)</sup> to 2<sup>(w-1)</sup>-1
 `long long`, `unsigned long long` | 1 to 64 | 0 to 2<sup>w</sup>-1
+
+
+
+# Function Calling Sequence
+This section describes the standard function calling sequence, including stack frame
+layout, Wasm argument and local value usage, and so on. These requirements apply
+only to global functions. Local functions not reachable from other compilation units
```
[the remark this answers is not in the document]
Comment (dschuff) [reply BasicCABI.md]: Clarified.

File: BasicCABI.md [hunk 2]
```diff
@@ -81,3 +81,58 @@ Bitfield type | Witdh *w* | Range
...
+and it points to the bottom of the static-size objects (those whose sizes are known at compile time). 
+If objects in the current frame require alignment greater than 16, then a base pointer (another local, `BP`) is used, which points to the bottom of the previous frame.
+The stack also has a "red zone" which extends 128 bytes below the stack pointer. If a function
+does not need a frame or base pointer, it may write data into the red zone which is not needed
+across function calls. So a leaf function which needs less than 128 bytes of stack space
+need not update the stack pointer in its prolog or epilog at all.
+
+The frame organization is illustrated as follows:
+
+Position | Contents | Frame
```
[the remark this answers is not in the document]
Comment (dschuff) [reply BasicCABI.md]: Natural to you, maybe :)
This diagram follows the convention used in the SysV ABI, although I wish markdown were better for tables.

[the remark this answers is not in the document]
Comment (binji) [reply BasicCABI.md]: Good point, though I feel like I usually see memory maps the other way. I don't have a ready example of this, of course :-}

[the remark this answers is not in the document]
Comment (dschuff) [reply BasicCABI.md]: I also just checked the AAPCS ABI and it agrees too, so I'll just leave it :)

**Comment by dschuff** [comment]
I'm going to go ahead and land this since it reflects the current reality in LLVM. We can propose changes (or doc improvements) in separate PRs.

File: BasicCABI.md [hunk 3]
```diff
@@ -81,3 +81,69 @@ Bitfield type | Witdh *w* | Range
...
+layout, Wasm argument and local value usage, and so on. These requirements apply
+only to global functions (those reachable from other compilation units) . Local functions
+may use different conventions; however this may reduce the ability of external tools
+to understand them.
+
+
+## Locals and the stack frame
+WebAssembly does not have registers in the style of hardware architectures. Instead it has an
+unlimited number of function
+arguments and local variables, which have wasm value types. These local values are generally
```
Comment (sbc100) [inline BasicCABI.md]: Can you reflow these paragraphs?

File: BasicCABI.md [hunk 4]
```diff
@@ -81,3 +81,69 @@ Bitfield type | Witdh *w* | Range
...
+The stack also has a "red zone" which extends 128 bytes below the stack pointer. If a function
+does not need a frame or base pointer, it may write data into the red zone which is not needed
+across function calls. So a leaf function which needs less than 128 bytes of stack space
+need not update the stack pointer in its prolog or epilog at all.
+
+The frame organization is illustrated as follows:
+
+Position | Contents | Frame
+-|-|-
+`BP` |  unspecified | Previous
```
Comment (sbc100) [inline BasicCABI.md]: Can you try make this table readable in the text format too?

**Pull request merged**